### ISSUE TITLE
Clippy is applying destructive changes within macros, breaking them

### ISSUE THREAD

**Issue by rscarson** [rust-lang/rust-clippy#12426]
### Summary

I managed to trigger an issue where clippy is poisoning macro definitions instead of fixing their invocations  
I triggered it with the `#redundant_field_names` lint

### Reproducer

I tried this code:

```rust
enum ErrorDetails {
    VariantA { value: u32 },

    VariantB { msg: String },
}

macro_rules! oops {
    ($variant:ident $({ $($n:ident$(:$v:expr)?),+ })?) => {
        ErrorDetails::$variant $({ $($n $(: $v)?),+ })?
    };
}

fn main() {
    let value = 42;
    oops!(VariantA { value: value });

    oops!(VariantB {
        msg: "Hello, World!".to_string()
    });
}
```

and run `cargo clippy --fix`

I expected to see this happen:

- remove the redundant `: value` in main

Instead, this happened:

```
after fixes were automatically applied the compiler reported errors within these files:

  * src\main.rs

This likely indicates a bug in either rustc or cargo itself,
and we would appreciate a bug report! You're likely to see
a number of compiler warnings after this message which cargo
attempted to fix but failed. If you could open an issue at
https://github.com/rust-lang/rust-clippy/issues
quoting the full output of this command we'd be very appreciative!
Note that you may be able to make some more progress in the near-term
fixing code with the `--broken-code` flag

The following errors were reported:
error[E0559]: variant `ErrorDetails::VariantB` has no field named `value`
  --> src\main.rs:9:38
   |
9  |           ErrorDetails::$variant $({ $(value $(: $v)?),+ })?
   |                                        ^^^^^ `ErrorDetails::VariantB` does not have this field
...
17 | /     oops!(VariantB {
18 | |         msg: "Hello, World!".to_string()
19 | |     });
   | |______- in this macro invocation
   |
   = note: all struct fields are already assigned
   = note: this error originates in the macro `oops` (in Nightly builds, run with -Z macro-backtrace for more info)

error: aborting due to 1 previous error
```

### Version

```text
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: x86_64-pc-windows-msvc
release: 1.76.0
LLVM version: 17.0.6
```

**Comment by PatchMixolydic** [comment]
#12407 is another instance of the same overarching issue, but with `clippy::manual_async_fn`.  (I haven't tried `cargo clippy --fix`, but [the erroneous suggestions are `MachineApplicable`](https://github.com/rust-lang/rust-clippy/blob/a79db2aa5173edcda81c7e571bd21d4097e8eb73/clippy_lints/src/manual_async_fn.rs#L88), so they should get automatically applied, as in this issue.)
